x-pack/plugins/security_solution/public/timelines/components/timeline/query_tab_content/index.tsx

```diff
@@ -0,0 +1,434 @@
+  }
+`;
+
+DatePicker.displayName = 'DatePicker';
+
+const VerticalRule = styled.div`
+  width: 2px;
+  height: 100%;
+  background: ${({ theme }) => theme.eui.euiColorLightShade};
+`;
+
+VerticalRule.displayName = 'VerticalRule';
+
+const StyledEuiTabbedContent = styled(EuiTabbedContent)`
+  display: flex;
+  flex-direction: column;
+  flex: 1;
+  overflow: hidden;
+
+  > [role='tabpanel'] {
+    display: flex;
+    flex-direction: column;
+    flex: 1;
+    overflow: hidden;
+  }
+`;
+
+StyledEuiTabbedContent.displayName = 'StyledEuiTabbedContent';
+
+const isTimerangeSame = (prevProps: Props, nextProps: Props) =>
+  prevProps.end === nextProps.end &&
+  prevProps.start === nextProps.start &&
+  prevProps.timerangeKind === nextProps.timerangeKind;
+
+interface OwnProps {
+  timelineId: string;
+}
+
+export type Props = OwnProps & PropsFromRedux;
+
+export const QueryTabContentComponent: React.FC<Props> = ({
+  columns,
+  dataProviders,
+  end,
+  eventType,
+  filters,
+  timelineId,
+  isLive,
+  itemsPerPage,
+  itemsPerPageOptions,
+  kqlMode,
+  kqlQueryExpression,
+  showCallOutUnauthorizedMsg,
+  showEventDetails,
+  start,
+  status,
+  sort,
+  timerangeKind,
+  updateEventTypeAndIndexesName,
+}) => {
+  const [showEventDetailsColumn, setShowEventDetailsColumn] = useState(false);
+
+  useEffect(() => {
+    // it should changed only once to true and then stay visible till the component umount
+    setShowEventDetailsColumn((current) => {
+      if (showEventDetails && !current) {
+        return true;
+      }
+      return current;
+    });
+  }, [showEventDetails]);
+
+  const {
+    browserFields,
+    docValueFields,
+    loading: loadingSourcerer,
+    indexPattern,
+    selectedPatterns,
+  } = useSourcererScope(SourcererScopeName.timeline);
+
+  const { uiSettings } = useKibana().services;
+  const [filterManager] = useState<FilterManager>(new FilterManager(uiSettings));
+  const esQueryConfig = useMemo(() => esQuery.getEsQueryConfig(uiSettings), [uiSettings]);
+  const kqlQuery = useMemo(() => ({ query: kqlQueryExpression, language: 'kuery' }), [
+    kqlQueryExpression,
+  ]);
+  const combinedQueries = useMemo(
+    () =>
+      combineQueries({
+        config: esQueryConfig,
+        dataProviders,
+        indexPattern,
+        browserFields,
+        filters,
+        kqlQuery,
+        kqlMode,
+      }),
+    [browserFields, dataProviders, esQueryConfig, filters, indexPattern, kqlMode, kqlQuery]
+  );
+
+  const canQueryTimeline = useMemo(
+    () =>
+      combinedQueries != null &&
+      loadingSourcerer != null &&
+      !loadingSourcerer &&
+      !isEmpty(start) &&
+      !isEmpty(end),
+    [loadingSourcerer, combinedQueries, start, end]
+  );
+
+  const columnsHeader = isEmpty(columns) ? defaultHeaders : columns;
+  const timelineQueryFields = useMemo(() => {
+    const columnFields = columnsHeader.map((c) => c.id);
+    return [...columnFields, ...requiredFieldsForActions];
+  }, [columnsHeader]);
```

Comment:
I think that we should this logic `columnsHeader ` inside of the memo and have the dependency on a props and not a on a local variables.